rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -227,32 +258,37 @@ class SignalHandler final
     }
 #endif
     signal_received.exchange(true);
-    events_condition_variable.notify_one();
+    RCLCPP_DEBUG(g_logger, "signal_handler(): SIGINT received, notifying deferred signal handler");
+    notify_signal_handler();
   }
 
   void
   deferred_signal_handler()
   {
     while (true) {
       if (signal_received.exchange(false)) {
+        RCLCPP_DEBUG(g_logger, "deferred_signal_handler(): SIGINT received, shutting down");
         for (auto context_ptr : rclcpp::get_contexts()) {
           if (context_ptr->get_init_options().shutdown_on_sigint) {
+            RCLCPP_DEBUG(
+              g_logger,
+              "deferred_signal_handler(): "
+              "shutting down rclcpp::Context @ %p, because it had shutdown_on_sigint == true",
+              context_ptr.get());
             context_ptr->shutdown("signal handler");
           }
         }
       }
       if (!is_installed()) {
+        RCLCPP_DEBUG(g_logger, "deferred_signal_handler(): signal handling uninstalled");
         break;
       }
-      std::unique_lock<std::mutex> events_lock(events_mutex);
-      events_condition_variable.wait(events_lock);
+      RCLCPP_DEBUG(g_logger, "deferred_signal_handler(): waiting for SIGINT or uninstall");
+      wait_for_signal();
```

Comment:
@wjwwood IIUC `wait_for_signal()` could fail. Should we log it / act upon it (e.g. forcing a shutdown)?

Reply:
I'll throw from here, which will cause the program to terminate, which I think is the best thing to do, as this function really _shouldn't_ fail, it just _can_.

Reply:
That's exactly the argument I was going to make on whether this function should be silent about errors or not. I'm not sure if just throwing will do though, AFAIK it'll kill this thread but not the rest. Maybe logging a fatal level error and going into shutdown or raising an abort signal?

Reply:
Throwing will kill it, since the exception is unhandled it will call `std::terminate()`.